core/modules/wcontrol/Foreman.cc

```diff
@@ -95,42 +97,61 @@ Foreman::~Foreman() {
     _pool->shutdownPool();
 }
 
-/// Put the task on the scheduler to be run later.
-void Foreman::processTask(std::shared_ptr<wbase::Task> const& task) {
 
+void Foreman::_setRunFunc(shared_ptr<wbase::Task> const& task) {
     auto func = [this, task](util::CmdData*){
         proto::TaskMsg const& msg = *task->msg;
         int const resultProtocol = 2; // See proto/worker.proto Result protocol
         if (!msg.has_protocol() || msg.protocol() < resultProtocol) {
             LOGS(_log, LOG_LVL_WARN, "processMsg Unsupported wire protocol");
-            if (!task->getCancelled()) {
+            if (!task->isCancelled()) {
                 // We should not send anything back to xrootd if the task has been cancelled.
+                mutex* streamMutex = task->sendChannel->getStreamMutexPtr();
+                lock_guard<mutex> streamLock(*streamMutex);
                 task->sendChannel->sendError("Unsupported wire protocol", 1);
             }
         } else {
             auto qr = wdb::QueryRunner::newQueryRunner(task, _chunkResourceMgr, _mySqlConfig,
-                                                       _sqlConnMgr, _transmitMgr);
+                    _sqlConnMgr, _transmitMgr);
             bool success = false;
             try {
                 success = qr->runQuery();
             } catch (UnsupportedError const& e) {
                 LOGS(_log, LOG_LVL_ERROR, "runQuery threw UnsupportedError " << e.what() << *task);
             }
             if (not success) {
+                mutex* streamMutex = task->sendChannel->getStreamMutexPtr();
```

Comment:
Please, see my earlier comment on preferring the "synchronized" methods in favor of exposing internal data structures (such as this mutex) to outside codes).